xmtp_mls/src/groups/sync.rs

```diff
@@ -0,0 +1,771 @@
+use super::{intents::SendMessageIntentData, members::GroupMember, GroupError, MlsGroup};
+use crate::{
+    api_client_wrapper::IdentityUpdate,
+    codecs::ContentCodec,
+    hpke::{encrypt_welcome, HpkeError},
+    retry_async,
+    storage::{
+        group_intent::{IntentKind, IntentState, ID},
+        refresh_state::EntityKind,
+        StorageError,
+    },
+    Fetch,
+};
+use log::debug;
+use openmls::{
+    framing::ProtocolMessage,
+    group::MergePendingCommitError,
+    prelude::{
+        LeafNodeIndex, MlsGroup as OpenMlsGroup, MlsMessageIn, MlsMessageInBody, PrivateMessageIn,
+        ProcessedMessage, ProcessedMessageContent, Sender,
+    },
+    prelude_test::KeyPackage,
+};
+use openmls_traits::OpenMlsProvider;
+use prost::Message;
+use std::{collections::HashMap, mem::discriminant};
+use tls_codec::{Deserialize, Serialize};
+use xmtp_proto::{
+    api_client::XmtpMlsClient,
+    xmtp::mls::api::v1::{
+        group_message::{Version as GroupMessageVersion, V1 as GroupMessageV1},
+        welcome_message_input::{
+            Version as WelcomeMessageInputVersion, V1 as WelcomeMessageInputV1,
+        },
+        GroupMessage, WelcomeMessageInput,
+    },
+    xmtp::mls::message_contents::GroupMembershipChanges,
+};
+
+use super::intents::{
+    AddMembersIntentData, AddressesOrInstallationIds, Installation, PostCommitAction,
+    RemoveMembersIntentData, SendWelcomesAction,
+};
+use crate::{
+    client::MessageProcessingError,
+    codecs::membership_change::GroupMembershipChangeCodec,
+    groups::validated_commit::ValidatedCommit,
+    identity::Identity,
+    retry,
+    retry::Retry,
+    storage::{
+        db_connection::DbConnection,
+        group_intent::StoredGroupIntent,
+        group_message::{GroupMessageKind, StoredGroupMessage},
+    },
+    utils::{hash::sha256, id::get_message_id},
+    xmtp_openmls_provider::XmtpOpenMlsProvider,
+    Delete, Store,
+};
+
+impl<'c, ApiClient> MlsGroup<'c, ApiClient>
+where
+    ApiClient: XmtpMlsClient,
+{
+    pub async fn sync(&self) -> Result<(), GroupError> {
+        let conn = &mut self.client.store.conn()?;
+        self.sync_with_conn(conn).await
+    }
+
+    pub(super) async fn sync_with_conn<'a>(
```

Comment:
`sync_with_conn` now continues even if there is a publishing error, so that the state can resolve itself by receiving messages from the network.